components/sup/src/manager.rs

```diff
@@ -1421,6 +1429,24 @@ impl Manager {
             event::service_stopped(&service);
             user_config_watcher.remove(&service);
             service_updater.lock().remove(&service.service_group);
+            // Remove any packages that are newer than the latest desired on restart package.
+            if let Some(latest_desired_ident) = latest_desired_on_restart {
+                while let Some(latest_installed) = pkg::installed(service.spec_ident.clone()) {
+                    let latest_ident = latest_installed.ident;
+                    if latest_ident > latest_desired_ident {
+                        info!("Track channel update condition is removing package '{}' inorder \
+                               to rollback to '{}'",
+                              latest_ident, latest_desired_ident);
+                        if let Err(e) = pkg::uninstall(&latest_ident).await {
+                            error!("Failed to uninstall '{}' during rollback. On restart, \
+                                    service will start with the wrong package. err: {}",
+                                   latest_ident, e);
+                        }
+                    } else {
+                        break;
+                    }
+                }
+            }
```

Comment:
This feels like a chunk of logic that could be extracted into a standalone function, rather than being embedded within the "stop service" function.

I'm also wondering if this logic should ultimately be invoked from the service updater threads themselves. They are the ones that are responsible for a) figuring out what the next version of the service to run should be and b) getting that version on disk in the first place. If you squint your eyes, "getting that version on disk in the first place" could straightforwardly become "ensuring that version is the _latest_ on disk", given the appropriate update condition. I think that would also prevent having to make a bunch of changes outside of the updaters themselves.

Reply:
> This feels like a chunk of logic that could be extracted into a standalone function

Addressed in [ff708a](https://github.com/habitat-sh/habitat/pull/7513/commits/ff708a4d80cdda8257d25a8e052fe6d64d3cef4c)

> I'm also wondering if this logic should ultimately be invoked from the service updater threads themselves.

Agreed, that is a much more intuitive way to think about it and this was the original approach I was taking. I ran into a dead-end though when the package had to be not "running" inorder to uninstall it. So I moved the logic here after the service is supposed to be not "running". Turns out "running" is only determined by if a spec file exists. So the service is always running and we cannot remove the spec file because we rely on the spec file in order to restart the service.

So we could move the uninstall logic back into the update workers, but then the uninstall will happen before these lines

```rust
service.stop_gsw(shutdown_config).await;
event::service_stopped(&service);
user_config_watcher.remove(&service);
service_updater.lock().remove(&service.service_group);
```

Seems like that might have the potential to cause some problems. Thoughts?

Ultimately, when updating and restarting is a core part of the service lifecycle this should be much cleaner.

Reply:
Yeah, deleting the package while the service is still running could cause headaches. In most cases it should probably be OK, since the binary is already in memory, but I'm sure there would be some package out there somewhere that does something "unconventional" that would trigger all kinds of weird behavior if we deleted before stopping.

Let's be conservative for now, and keep an eye out for future cleanups as we continue to formalize the service lifecycle.